### PR TITLE
Add TPM NV extend support. 

### PR DESCRIPTION
**Pull request by qizhangz**

### CODE REVIEW

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c [hunk 1]
```diff
@@ -16,6 +16,8 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Library/MemoryAllocationLib.h>
 #include <Library/PcdLib.h>
 #include <Library/HashLib.h>
+#include <Protocol/Tcg2Protocol.h>
+#include <Protocol/TrEEProtocol.h>
```
Comment (jyao1) [inline SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c]: why need this?

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c [hunk 2]
```diff
@@ -16,6 +16,8 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Library/MemoryAllocationLib.h>
 #include <Library/PcdLib.h>
 #include <Library/HashLib.h>
+#include <Protocol/Tcg2Protocol.h>
```
Comment (jyao1) [inline SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c:19]: why need this?

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c [hunk 3]
```diff
@@ -164,21 +199,55 @@ HashCompleteAndExtend (
...
+
+    for (Index = 0; Index < mHashInterfaceCount; Index++) {
+      HashMask = Tpm2GetHashMaskFromAlgo (&mHashInterface[Index].HashGuid);
+      AlgoId = Tpm2GetAlgoIdFromAlgo (&mHashInterface[Index].HashGuid);
+      if (((HashMask & PcdGet32 (PcdTpm2HashMask)) != 0)
+           && (AlgoId == PublicInfo.nvPublic.nameAlg)) {
+        mHashInterface[Index].HashUpdate (HashCtx[Index], DataToHash, DataToHashLen);
+        mHashInterface[Index].HashFinal (HashCtx[Index], &Digest);
+        Tpm2SetHashToDigestList (DigestList, &Digest);
+      }
```
Comment (jyao1) [inline SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c:240]: what if no match?

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c [hunk 4]
```diff
@@ -164,21 +199,55 @@ HashCompleteAndExtend (
...
+    Status = Tpm2NvReadPublic (
+               PcrIndex,
+               &PublicInfo,
+               &PubName
+               );
+    if (EFI_ERROR(Status)) {
+      DEBUG ((DEBUG_ERROR, "TPM2 Nv 0x%x ReadPublic failed %r\n", PcrIndex, Status));
+      return Status;
+    }
+    DataSize = PublicInfo.nvPublic.dataSize;
```
Comment (jyao1) [inline SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c:228]: please assert DataSize==HashSize

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c (hunk 1 above)
Comment (qizhangz) [reply SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c]: MAX_PCR_INDEX is defined in these two header files.

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c (hunk 2 above)
Comment (qizhangz) [reply SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c:19]: MAX_PCR_INDEX is defined in these two header files.

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c (hunk 3 above)
Comment (qizhangz) [reply SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c:240]: add the below check if no match.
    if (DigestList->count == 0) {
      return EFI_NOT_FOUND;
    }

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c (hunk 4 above)
Comment (qizhangz) [reply SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c:228]: done

File: SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c [hunk 5]
```diff
@@ -154,6 +185,11 @@ HashCompleteAndExtend (
   UINTN               Index;
   EFI_STATUS          Status;
   UINT32              HashMask;
+  TPM2B_NV_PUBLIC    PublicInfo;
```
Comment (jyao1) [inline SecurityPkg/Library/HashLibBaseCryptoRouter/HashLibBaseCryptoRouterDxe.c]: this is not aligned.

**Review by jyao1 — APPROVED** [review]

**Pull request merged**